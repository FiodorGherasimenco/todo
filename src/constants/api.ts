export const TASKS_URLS = "https://dummyjson.com";
export const DEFAULT_USER_ID = 1;
