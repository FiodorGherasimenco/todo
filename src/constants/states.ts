export const Idle = "idle";
export const Create = "create";
export const Fetch = "fetch";
export const Error = "error";
export const Change = "change";
export const Resert = "reset";
export const Complete = "complete";
export const Delete = "delete";
export const Hide = "hide";
