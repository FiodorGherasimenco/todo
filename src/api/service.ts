import { TASKS_URLS } from "../constants/api";

export const service = <TOutput>(
  path: string,
  config?: RequestInit,
): Promise<TOutput> =>
  fetch(new URL(path, TASKS_URLS), {
    ...config,
    headers: {
      ...config?.headers,
      "Content-Type": "application/json",
    },
  }).then((response) => {
    if (response.status > 400) {
      throw new Error("Something went wrong");
    }

    return response.json();
  });
