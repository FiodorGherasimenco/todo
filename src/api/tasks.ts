import { DEFAULT_USER_ID } from "../constants/api";
import { service } from "./service";

export type Todo<T = object> = T & {
  id: number;
  todo: string;
  completed: boolean;
};

export type TodosOutput<T = object> = {
  todos: Todo<T>[];
  total: number;
  skip: number;
  limit: number;
};

type TodosConfig = {
  skip?: number;
  limit?: number;
};

export const getAllTasks = ({ skip = 0, limit = 10 }: TodosConfig = {}) =>
  service<TodosOutput>(
    `/todos?${new URLSearchParams({
      skip: String(skip),
      limit: String(limit),
    }).toString()}`,
  );

export const addTask = (todo: string) =>
  service<Todo>("/todos/add", {
    method: "POST",
    body: JSON.stringify({
      todo: todo,
      completed: false,
      userId: DEFAULT_USER_ID,
    }),
  });

export const completeTask = (taskId: number) =>
  service<Todo>(`/todos/${taskId}`, {
    method: "PUT",
    body: JSON.stringify({
      completed: true,
    }),
  });

export const deleteTask = (taskId: number) =>
  service<
    Todo<{
      isDeleted: boolean;
      deletedOn: string;
    }>
  >(`/todos/${taskId}`, {
    method: "DELETE",
    body: JSON.stringify({
      completed: true,
    }),
  });
