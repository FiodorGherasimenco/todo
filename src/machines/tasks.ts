import * as States from "../constants/states";
import { createMachine } from "../fsm";
import { getAllTasks, Todo, addTask } from "../api/tasks";

type Context = { tasks: Todo[] };

let timeout: NodeJS.Timeout;

export default createMachine<Context>({
  initialState: States.Idle,
  context: {
    tasks: [],
  },
  states: {
    [States.Idle]: {
      transitions: {
        fetch: {
          target: States.Fetch,
        },
        add: {
          target: States.Create,
        },
      },
    },
    [States.Fetch]: {
      actions: {
        onEnter(_, instance) {
          getAllTasks()
            .then(({ todos }) => {
              instance.setContext((context) => ({
                ...context,
                tasks: todos,
              }));
              instance.transition("complete");
            })
            .catch(() => {
              instance.transition("error");
            });
        },
      },
      transitions: {
        complete: {
          target: States.Idle,
        },
        error: {
          target: States.Error,
        },
      },
    },
    [States.Create]: {
      actions: {
        onEnter({ title }, instance) {
          if (!title || typeof title !== "string" || !title.length) {
            instance.transition("complete");
          } else {
            addTask(title)
              .then((task) => {
                instance.setContext((context) => ({
                  ...context,
                  tasks: [...context.tasks, task],
                }));
                instance.transition("complete");
              })
              .catch(() => {
                instance.transition("error");
              });
          }
        },
      },
      transitions: {
        complete: {
          target: States.Idle,
        },
        error: {
          target: States.Error,
        },
      },
    },
    [States.Error]: {
      actions: {
        onEnter(_, instance) {
          timeout = setTimeout(() => {
            instance.transition("switch");
          }, 3000);
        },
        onExit() {
          clearTimeout(timeout);
        },
      },
      transitions: {
        reset: {
          target: States.Idle,
        },
      },
    },
  },
});
