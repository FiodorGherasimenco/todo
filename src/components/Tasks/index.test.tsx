import {
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import { http, HttpResponse } from "msw";
import { setupServer } from "msw/node";
import { beforeAll, beforeEach, describe, expect, test } from "vitest";
import { Todo, TodosOutput } from "../../api/tasks";
import { TASKS_URLS } from "../../constants/api";
import { Tasks } from "./index";

const server = setupServer();

describe("<Tasks>", () => {
  beforeAll(() => {
    server.listen();
  });

  beforeEach(() => {
    cleanup();
  });

  test("snapshot", async () => {
    server.use(
      http.get(`${TASKS_URLS}/todos`, () =>
        HttpResponse.json<TodosOutput>({
          todos: [
            {
              id: 1,
              todo: "fake_title",
              completed: false,
            },
          ],
          total: 1,
          skip: 0,
          limit: 10,
        }),
      ),
    );

    const { container } = render(<Tasks />);
    expect(container).toMatchSnapshot();

    await waitFor(() => screen.getByRole("button", { name: /delete/i }));

    expect(container).toMatchSnapshot();
  });

  test("add new task", async () => {
    server.use(
      http.get(`${TASKS_URLS}/todos`, () =>
        HttpResponse.json<TodosOutput>({
          todos: [],
          total: 0,
          skip: 0,
          limit: 10,
        }),
      ),
      http.post<never, Omit<Todo, "id">, Todo>(
        `${TASKS_URLS}/todos/add`,
        async ({ request }) => {
          const json = await request.json();
          return HttpResponse.json<Todo>({
            id: 1,
            ...json,
          });
        },
      ),
    );

    render(<Tasks />);

    const buttonCreate = screen.getByRole("button", {
      name: /create/i,
    });
    const inputText = screen.getByRole("textbox");

    expect(inputText.hasAttribute("disabled")).toBeTruthy();
    expect(buttonCreate.hasAttribute("disabled")).toBeTruthy();

    await waitFor(() => expect(inputText.hasAttribute("disabled")).toBeFalsy());

    fireEvent.input(inputText, { target: { value: "dummy test task" } });

    expect(buttonCreate.hasAttribute("disabled")).toBeFalsy();

    fireEvent.click(buttonCreate);

    await waitFor(() =>
      expect(screen.getByText("dummy test task")).toBeDefined(),
    );
  });
});
