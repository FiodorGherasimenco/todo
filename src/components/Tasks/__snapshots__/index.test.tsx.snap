// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Tasks> > snapshot 1`] = `
<div>
  <div
    class="_container_f7bf23"
  >
    <form
      class="_form_adff70"
    >
      <input
        class="_input_4f1997 _medium_4f1997 _fluid_4f1997"
        disabled=""
        value=""
      />
      <button
        class="_button_0696ee _primary_0696ee _medium_0696ee"
        disabled=""
        type="submit"
      >
        Create
      </button>
    </form>
    <div
      class="_tasks_f7bf23"
    >
      <div
        class=""
      >
        Loading...
      </div>
    </div>
  </div>
</div>
`;

exports[`<Tasks> > snapshot 2`] = `
<div>
  <div
    class="_container_f7bf23"
  >
    <form
      class="_form_adff70"
    >
      <input
        class="_input_4f1997 _medium_4f1997 _fluid_4f1997"
        value=""
      />
      <button
        class="_button_0696ee _primary_0696ee _medium_0696ee"
        disabled=""
        type="submit"
      >
        Create
      </button>
    </form>
    <div
      class="_tasks_f7bf23"
    >
      <div
        class="_task_78c96f"
      >
        <div
          class="_truncated_78c96f false"
        >
          fake_title
        </div>
        <div
          class="_actions_78c96f"
        >
          <button
            class="_button_0696ee _secondary_0696ee _medium_0696ee"
          >
            Complete
          </button>
          <button
            class="_button_0696ee _secondary_0696ee _medium_0696ee"
          >
            Delete
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
